cirq/optimizers/controlled_gate_decomposition.py

```diff
@@ -0,0 +1,254 @@
+        return [*part1, *part2, *part1, *part2]
+    else:
+        # No free qubits - must use general algorithm.
+        # This will never happen if called from main algorithm and is added
+        # only for completeness.
+        return decompose_multi_controlled_unitary(unitary(ops.X), controls,
+                                                  target)
+
+
+def _decompose_su(matrix: np.ndarray, controls: List['cirq.Qid'],
+                  target: 'cirq.Qid') -> List['cirq.Operation']:
+    """Decomposes controlled special unitary gate into elementary gates.
+
+    Result has O(len(controls)) operations.
+    See [1], lemma 7.9.
+    """
+    assert matrix.shape == (2, 2)
+    assert is_special_unitary(matrix)
+    assert len(controls) >= 1
+
+    a, b, c, _ = _decompose_abc(matrix)
+
+    cnots = decompose_multi_controlled_x(controls[:-1], target, [controls[-1]])
+    return [
+        *_decompose_single_ctrl(c, controls[-1], target), *cnots,
+        *_decompose_single_ctrl(b, controls[-1], target), *cnots,
+        *_decompose_single_ctrl(a, controls[-1], target)
+    ]
+
+
+def _decompose_recursive(matrix: np.ndarray, power: float,
+                         controls: List['cirq.Qid'], target: 'cirq.Qid',
+                         free_qubits: List['cirq.Qid']
+                        ) -> List['cirq.Operation']:
+    """Decomposes controlled unitary gate into elementary gates.
+
+    Result has O(len(controls)^2) operations.
+    See [1], lemma 7.5.
+    """
+    if len(controls) == 1:
+        return _decompose_single_ctrl(_unitary_power(matrix, power),
+                                      controls[0], target)
+
+    cnots = decompose_multi_controlled_x(controls[:-1], controls[-1],
+                                         free_qubits + [target])
+    return [
+        *_decompose_single_ctrl(_unitary_power(matrix, 0.5 * power),
+                                controls[-1], target), *cnots,
+        *_decompose_single_ctrl(_unitary_power(matrix, -0.5 * power),
+                                controls[-1], target), *cnots,
+        *_decompose_recursive(matrix, 0.5 * power, controls[:-1], target,
+                              [controls[-1]] + free_qubits)
+    ]
+
+
+def decompose_multi_controlled_unitary(matrix: np.ndarray,
+                                       controls: List['cirq.Qid'],
+                                       target: 'cirq.Qid'
+                                      ) -> List['cirq.Operation']:
+    """Implements action of multi-controlled unitary gate.
+
+    Returns a sequence of operations, which is equivalent to applying
+    single-qubit gate with matrix `matrix` on `target`, controlled by
+    `controls`.
+
+    Result is guaranteed to consist exclusively of 1-qubit, CNOT and CCNOT
+    gates.
+
+    If matrix is special unitary, result has length `O(len(controls))`.
+    Otherwise result has length `O(len(controls)**2)`.
+
+    === REFERENCE ===
```

Comment:
Make this a `References:` section similar to `Args:`.

Reply:
Done.